edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,747 @@
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(nameof(this.CreateOrUpdateModuleAsync), ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, this.FormatErrorResponseMessage(ex.Message));
+                    return;
+                }
+
+                IHttpRequestAuthenticator authenticator = await this.authenticatorGetter;
+                HttpAuthResult authResult = await authenticator.AuthenticateAsync(deviceId, Option.None<string>(), this.HttpContext);
+
+                if (!authResult.Authenticated)
+                {
+                    Events.AuthenticateFail(nameof(this.CreateOrUpdateModuleAsync), deviceId);
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                Events.Authenticated(nameof(this.CreateOrUpdateModuleAsync), deviceId);
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                string targetId = $"{deviceId}/{moduleId}";
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+
+                (bool authorized, string authChain) = await this.AuthorizeActorAsync(identitiesCache, edgeDeviceId, Constants.EdgeHubModuleId, targetId);
+                if (!authorized)
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                Events.Authorized(nameof(this.CreateOrUpdateModuleAsync), edgeDeviceId, Constants.EdgeHubModuleId, targetId, authChain);
+                var requestData = new CreateOrUpdateModuleOnBehalfOfData { AuthChain = $"{AuthChainHelpers.SkipFirstIdentityFromAuthChain(authChain)}", Module = module };
+                HttpResponseMessage responseMessage = await this.apiClient.PutModuleOnBehalfOfAsync(edgeDeviceId, requestData);
+                await this.SendResponseAsync(responseMessage);
+                Events.CompleteRequest(nameof(this.CreateOrUpdateModuleAsync), edgeDeviceId, targetId, requestData.AuthChain);
+            }
+            catch (Exception ex)
+            {
+                Events.InternalServerError(nameof(this.CreateOrUpdateModuleAsync), ex);
+                await this.SendResponseAsync(HttpStatusCode.InternalServerError, this.FormatErrorResponseMessage(ex));
+            }
+        }
+
+        [HttpPut]
+        [Route("devices/{actorDeviceId}/putModuleOnBehalfOf")]
+        public async Task CreateOrUpdateModuleOnBehalfOfAsync(
+            [FromRoute] string actorDeviceId,
+            [FromBody] CreateOrUpdateModuleOnBehalfOfData requestData)
+        {
+            try
+            {
+                Events.ReceivedOnBehalfOfRequest(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), actorDeviceId, Events.GetAdditionalInfo(requestData));
```

Comment:
Lot of duplicate code in these methods. Maybe we should refactor?

Reply:
Would you be specific what kind of refactoring do you mean? if using func delegate, I would like to keep it simple and don't change now.